File: SampleApps /ObjC/SampleAppWakeUp/SampleAppWakeUp/UI/ViewController.m
//
//  ViewController.m
//  SampleAppWakeUp
//
//  Created by Akhil Rana on 24/11/23.
//

#import "ViewController.h"
#import "Constants.h"
#import "Notification.h"
#import "MistService.h"
#import "WakeupService.h"

@interface ViewController () <MistServiceDelegate, WakeUpServiceDelegate>
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UIButton *wakeUpButton;

@property (strong, nonatomic) MistService *mistService;
@property (strong, nonatomic) WakeupService *wakeupService;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setup];
 }

- (void)setup {
    self.mistService = [[MistService alloc] initWithToken: SDK_TOKEN];
    self.mistService.delegate = self;
    self.wakeupService = [[WakeupService alloc] init];
    self.wakeupService.delegate = self;
    _statusLabel.text = notMonitoringMessage;
}

- (IBAction)toggleWakeUpStatus:(id)sender {
    if (_wakeupService.isWakeUpRunning) {
        [_wakeUpButton setTitle:@"Start WakeUp Service" forState: UIControlStateNormal];
        _statusLabel.text = notMonitoringMessage;
        [_wakeupService stop];
    } else {
        [_wakeUpButton setTitle:@"Stop WakeUp Service" forState: UIControlStateNormal];
        _statusLabel.text = monitoringMessage;
        [_wakeupService start];
    }
    _wakeupService.isWakeUpRunning = !_wakeupService.isWakeUpRunning;
}

- (void) didEnterRegion:(NSString *)region {
    [self startMistServiceWithNotification:region];
}

- (void) didExitRegion:(NSString *)region {
    [Notification scheduleLocalNotificationWithTitle:@"Thankyou!!!" subTitle:region body:@"Welcome again!"];
 }

- (void) startMistServiceWithNotification: (NSString *)region {
    NSLog(@">>> didEnterRegion Beacon = %@", region);
    
    [Notification scheduleLocalNotificationWithTitle:@"Hello!!!" subTitle:region body:@"Welcome!! You have entered the Beacon range"];
    
    if (_mistService.isMistRunning == NO) {
        // Start Mist SDK
        [_mistService start];
        
        [Notification scheduleLocalNotificationWithTitle:@"Mist Service Started!" subTitle:@"" body: @"Mist Location Tracking Running..."];
    }
}

@end

File: Sample Apps /SampleAppWakeUp/SampleAppWakeUp/ViewController.swift
//
//  ViewController.swift
//  SampleAppWakeUp
//
//  Created by Rajesh Vishwakarma on 27/01/23.
//

import UIKit
import CoreLocation

protocol ViewDelegate: AnyObject {
    func didUpdateRelativeLocation(_ location: CGPoint)
}

class ViewController: UIViewController {
    
    @IBOutlet var messageLabel: UILabel!
    @IBOutlet var wakeUpButton: UIButton!
        
    var viewModel: ViewModelWakeUpDelegate?
        
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        configureViewModel()
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
        messageLabel.text = Mist.WakeUp.notMonitoringMessage
        
        wakeUpButton.isSelected = false
        wakeUpButton.setTitle("Start WakeUp Service", for: .normal)
        wakeUpButton.setTitle("Stop WakeUp Service", for: .selected)
        updateStatus()
    }
    
    private func configureViewModel() {
        #if !targetEnvironment(simulator)
        let wakeUpService = RealWakeupService()
        let mistService = RealMistService(token: Mist.SDK.token)
        let viewModel = ViewModel(wakeUpService: wakeUpService, mistService: mistService)
        viewModel.viewDelegate = self
        wakeUpService.delegate = viewModel
        mistService.delegate = viewModel
        self.viewModel = viewModel
        #endif
    }

    @IBAction func toggleWakeUpStatus() {
        guard let viewModel = viewModel else { return }
                
        if viewModel.isWakeUpRuning {
            viewModel.stopWakeUp()
        } else {
            viewModel.startWakeUp()
        }
    
        updateStatus()
    }
    
    func updateStatus() {
        guard let viewModel = viewModel else { return }
        wakeUpButton.isSelected = viewModel.isWakeUpRuning
        messageLabel.text = wakeUpButton.isSelected ? Mist.WakeUp.monitoringMessage : Mist.WakeUp.notMonitoringMessage
    }
}

extension ViewController: ViewDelegate {
        
    func didUpdateRelativeLocation(_ location: CGPoint) {
        debugPrint(">>> didUpdateRelativeLocation = x=\(location.x) y=\(location.y)")
    }
}
